crates/ra_ide_api/src/rename.rs

```diff
@@ -0,0 +1,136 @@
+use relative_path::RelativePathBuf;
+
+use hir::{
+    self, ModuleSource, source_binder::module_from_declaration,
+};
+use ra_syntax::{
+    algo::find_node_at_offset,
+    ast,
+    AstNode,
+    SyntaxNode
+};
+
+use crate::{
+    db::RootDatabase,
+    FilePosition,
+    FileSystemEdit,
+    SourceChange,
+    SourceFileEdit,
+};
+use ra_db::{FilesDatabase, SyntaxDatabase};
+use relative_path::RelativePath;
+
+pub(crate) fn rename(
+    db: &RootDatabase,
+    position: FilePosition,
+    new_name: &str,
+) -> Option<SourceChange> {
+    let source_file = db.source_file(position.file_id);
+    let syntax = source_file.syntax();
+
+    if let Some((ast_name, ast_module)) = find_name_and_module_at_offset(syntax, position) {
+        rename_mod(db, ast_name, ast_module, position, new_name)
+    } else {
+        rename_reference(db, position, new_name)
+    }
+}
+
+fn find_name_and_module_at_offset(
+    syntax: &SyntaxNode,
+    position: FilePosition,
+) -> Option<(&ast::Name, &ast::Module)> {
+    let ast_name = find_node_at_offset::<ast::Name>(syntax, position.offset);
+    let ast_name_parent = ast_name
+        .and_then(|n| n.syntax().parent())
+        .and_then(|p| ast::Module::cast(p));
```

Comment:
We can use `?` instead of `and_then`s here

Reply:
forget it.